src/commands/commands.ts

```diff
@@ -2,28 +2,111 @@ import {Command, flags} from '@oclif/command'
 import ux from 'cli-ux'
 import * as _ from 'lodash'
 
+type Dictionary = {[index: string]: object}
 export default class Commands extends Command {
   static description = 'list all the commands'
 
   static flags = {
     help: flags.help({char: 'h'}),
-    json: flags.boolean({char: 'j', description: 'output in json format'}),
-    hidden: flags.boolean({description: 'also show hidden commands'}),
+    json: flags.boolean({char: 'j', description: 'show detailed command information in json format'}),
+    hidden: flags.boolean({description: 'show hidden commands'}),
+    verbose: flags.boolean({
+      description: 'include extended command information; must be used with --json',
+      dependsOn: ['json'],
+    }),
+    topic: flags.string({description: 'filter on topic'}),
+    plugin: flags.string({description: 'filter on plugin'}),
+    command: flags.string({description: 'filter on command name with a regular expression'}),
   }
 
   async run() {
     const {flags} = this.parse(Commands)
-    let commands = this.config.commands
+    let commands = this.getCommands()
     if (!flags.hidden) {
       commands = commands.filter(c => !c.hidden)
     }
+
     commands = _.sortBy(commands, 'id')
+
+    if (flags.topic) {
+      commands = _.filter(commands, command => command.id.startsWith(`${flags.topic}:`))
+    }
+
+    if (flags.plugin) {
+      commands = _.filter(commands, command => command.pluginName === flags.plugin)
+    }
+
+    if (flags.command) {
+      commands = _.filter(commands, command => new RegExp(flags.command || '').test(command.id))
+    }
+
     if (flags.json) {
-      ux.styledJSON(commands)
+      if (flags.verbose) {
+        ux.styledJSON(commands.map(cmd => {
+          let commandClass = cmd.load()
+          const obj = Object.assign({}, cmd, commandClass)
+
+          // Load all properties on all extending classes.
```

Comment:
I'm all ears if there is a better way to do this. The only other option I could think of was to have a `static toJson` method on the extended command class, but it would be annoying for a command implementer to have to specify all the properties on the command level and duplicate that in the toJson method.